xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift

```diff
@@ -288,61 +287,85 @@ struct DetailStackModel: Hashable, ModelProtocol {
         return Update(state: model)
     }
     
-    static func rebaseLinkOnAddress(
-        address: Slashlink,
-        link: SubSlashlinkLink
-    ) -> Slashlink {
-        // Stitch the base address on to the tapped link, making any
-        // bare slashlinks relative to the sphere they belong to.
-        //
-        // This is needed in the viewer but address will always based
-        // on our sphere in the editor case.
-        guard case let .petname(basePetname) = address.peer else {
-            return link.slashlink
+    public static func findBestAddressForLink(
+        context: Peer?,
+        link: SubSlashlinkLink,
+        environment: TestableEnvironment
+    ) async throws -> Slashlink {
+        let slashlink = link.slashlink.rebaseIfNeeded(peer: context)
+        let petname = link.slashlink.petname
+        let ourIdentity = try await environment.noosphere.identity()
+        
+        // We want the find the DID of this user so we can check if we follow them.
+        // If we do follow them, we should prefer our petname for them when navigating.
+        
+        // i.e. I am following @bob, @alice and @charlie
+        // I am viewing @bob's note /hello at @bob/hello
+        // There is a link in @bob's note to @charlie.alice/hey
+        // The relative address would be @charlie.alice.bob/hey
+        // BUT if we resolve the address we realise that we know @charlie already!
+        // So we rewrite the address to @charlie/hey
+        
+        // We _could_ also choose to simply bail out and navigate to the address without
+        // traversing, at the expense of the clever redirect.
+        var did: Did? = nil
+        if let petname = slashlink.petname {
+            do {
+                // Any errors during traversal mean we should give up
+                let sphere = try await environment.noosphere.traverse(petname: petname)
+                did = try await sphere.identity()
+            } catch {
+                did = nil
+            }
+        }
```

Comment:
Nit: this could be rewritten in expression form as

```swift
let did = try? await Func.run {
    let sphere = try await environment.noosphere.traverse(petname: petname)
    return try await sphere.identity()
}
```

I find expression-form code easier to follow when possible.

No action required.